storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.storage;
+
+import com.google.common.base.Throwables;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocalFileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.stream.Stream;
+
+/**
+ * :: Unstable ::
+ * <p>
+ * The [[LogStore]] implementation for GCS, which uses gcs-connector to
+ * provide the necessary atomic and durability guarantees:
+ *
+ * <ol>
+ *   <li>Atomic Visibility: Read/read-after-metadata-update/delete are strongly
+ * consistent for GCS.</li>
+ *
+ *   <li>Consistent Listing: GCS guarantees strong consistency for both object and
+ * bucket listing operations.
+ * https://cloud.google.com/storage/docs/consistency</li>
+ *
+ *   <li>Mutual Exclusion: Preconditions are used to handle race conditions.</li>
+ * </ol>
+ * Regarding file creation, this implementation:
+ * <ul>
+ *    <li>Opens a stream to write to GCS otherwise.</li>
+ *    <li>Throws [[FileAlreadyExistsException]] if file exists and overwrite is false.</li>
+ *    <li>Assumes file writing to be all-or-nothing, irrespective of overwrite option.</li>
+ * </ul>
+ *
+ * @note This class is not meant for direct access but for configuration based on storage system.
+ * See https://docs.delta.io/latest/delta-storage.html for details.
+ */
+@InterfaceStability.Unstable
+public class GCSLogStore extends HadoopFileSystemLogStore {
+
+    String preconditionFailedExceptionMessage = "412 Precondition Failed";
+
+    public GCSLogStore(Configuration hadoopConf) {
+        super(hadoopConf);
+    }
+
+    private static <T> T runInNewThread(
+            String threadName,
+            Boolean isDaemon,
+            Callable<T> body) throws Exception {
+        List<Exception> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
```

Comment:
why not `T result = null`

Reply:
T result = null, cannot be accessed within the inner class, it needs to be marked final to make it accessible. So created a placeholder ArrayList.

Reply:
sounds good. add an inline doc explaining why you are doing it this way then.